Array 1.41.0 announcement (#4512)
PostHog/posthog.com: contents/blog/the-posthog-array-1-41-0.md

---
date: 2022-10-31
title: "Array 1.41.0: Enable hedgehog mode"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
categories: ['Release notes', 'Product updates']
featuredImage: ../images/blog/posthog-array-blog.png
featuredImageType: full
---

Want to know more about what we're up to? [Subscribe to HogMail, our newsletter](/newsletter), which we send every two weeks!

> Running a self-hosted instance? Check out our [guide to upgrading PostHog](/docs/runbook/upgrading-posthog).

> **IMPORTANT NOTE FOR SELF-HOSTED USERS:** Self-hosted users should check `alias` usage before [upgrading PostHog](/docs/runbook/upgrading-posthog) due to changes involving persons on events. For example, assuming `email` is used as the identified user id, then: 

```
identify(email)                                   # in the frontend
alias(backend_unique_id, email)      # in the backend - THIS WILL NOT WORK
alias(email, backend_unique_id2)    # in the backend - the right order
```

> After updating to 1.41.0, check [ingestion warnings](#new-ingestion-warnings) and run async migrations 0005, 0006 and 0007.

Reviewer: This is also a bit more complex. Essentially the guide is:

First check ingestion warnings & fix everything there.
Note that if you're using alias in the wrong order your users will not be merged anymore from those bad calls, so if fixing this will take a long time you might want to downgrade to 1.40 - not sure if want to call this out or not 🤷‍♀️ 

Then
# if you haven't ran 0007 (and potentially not 0005, 0006 either) async migration:
Run the async migrations if you haven't ran them yet, they are required to be completed before upgrading to the next version, you can do so on the async migrations UI at `<your-posthog-site>/instance/async_migrations`.

Completing async migration 0007 enables persons-on-events. 
You can disable it from instance settings `PERSON_ON_EVENTS_ENABLED` if your queries aren't working properly - this can be mentioned in the section we have for persons-on-events.

# if async migration 0007 is completed 

IMPORTANT: If you either changed anything about alias or based of ingestion warnings or are using group you will need to re-run 0007 (make sure to do this after fixing all the problems from ingestion warnings).

How to re-run 0007:
1. connect to postgres pod: https://posthog.com/docs/self-host/deploy/troubleshooting#how-do-i-connect-to-postgres
2. update the 0007 async migration to be Not Started (it's safe it was in Completed state before), by running:
```
UPDATE posthog_asyncmigration SET status = 0 WHERE name = '0007_persons_and_groups_on_events_backfill' AND status = 2;
```
3. click on the run button on the async migrations UI at `<your-posthog-site>/instance/async_migrations`

Author: Thanks Tiina - I've made some updates.

- New: Persons on events on by default
- New: Count of events per user
- New: Text cards on dashboards
- New: App metrics
- New: Ingestion warnings
- New: Change your email yourself
- New: Hedgehog mode
- Improved: Recordings interface
- One more thing: Site apps 

### New: Persons on events on by default

We used to store events in one table and persons in another table. That meant that, once you reached billion event scale, any query which touched person properties would time out. But, no longer! After running [an extensive beta since update 1.39.0](/blog/the-posthog-array-1-39-0#beta-improving-query-performance-by-combining-persons-and-events), we've now moved person data onto the events themselves. This change also applies to groups, adding group properties and aggregation to the same table. 

You won’t see any UI changes as a result of this change — persons will still have their own Persons & Groups section on the sidebar, for example — but you will notice results are a lot (up to 400%!) faster for any queries involving persons, groups or events. This is a massive change, so be sure to [read the full announcement for more info](/blog/persons-on-events).

> > **Note:** As a result of putting persons on events we now have to be strict about identified and unidentified users. Now, only anonymous users can be aliased into other users - meaning that once you call `identify` on a user, you can no longer alias the user into another identified user. Because of that, we recommend double checking your `identify` and `alias` calls, and following advice above [how to upgrade a self-hosted instance to 1.41.0](/blog/the-posthog-array-1-41-0). 

Reviewer: for alias changes check out this email that we sent to the users https://docs.google.com/document/d/1qbd52Mmf8DgOPbFB4bs2vAHNi1gSPNw2kGxPD4Khewk/edit & my outdated comment.

Reviewer: @joethreepwood  lets just make it be short and simple here:

```suggestion
> **Note:** As a result of adding persons on events and changing how queries work we now have to be more strict about merging users. Because of that, we recommend double checking your `alias` calls, and following advice above [how to upgrade a self-hosted instance to 1.41.0](/blog/the-posthog-array-1-41-0). 
```

### New: Count of events per user

![count events per user](../images/blog/array/1-41-0-count-events-per-user.png)

Have you ever asked yourself '*What's the average average number of purchases per user?*' or '*What's the maximum number of forms submitted per user?'* 

Questions like these used to be hard to answer with PostHog, but no more! We've added a new series option for insights, enabling you to see the count of events per user - including sub-options for seeing the averages, minimum and maximum count of events per user. 

**Bonus update:** While working on this feature, we also took the opportunity to improve our math selector UI.

### New: Text cards on dashboards

> ![text cards on dashboards](../images/blog/array/1-41-0-text-card-on-dashboard.png)

Reviewer: I'd use a pie chart there instead of trend as we can't tell who's a yes and who's a no on it & maybe mention the source is our users i.e. https://posthog.com/404

We think session recording feels like an entirely new experience now, so check it out. Now's the perfect time to explore [the console log beta too](/manual/recordings#console-logs-recording-beta)!

### New: App metrics

PICTURE GOES HERE

Curious how well your plugins are doing? You no longer have to read smoke signals in AWS logs - just use the new Plugin Metrics page. 

### New: Persons on events on by default

PICTURE GOES HERE

We used to store events in one table and persons in another. Any query that touched person properties would require 
database level JOINs and time out, once you reached the billion event scale. No longer! By moving person data
onto the events themselves, we eliminated a whole class of performance issues 🚀.

### New: Alias changes

PICTURE GOES HERE

The above change required us to be strict about who's a identified user and who's not. Because of that, you might
benefit from double checking your `identify` and `alias` calls. The new rule is this: only anonymous users can be
aliased into other users. Once you call `identify` on a user, you can no longer alias this user into another identified

> user. 

Author: @tiina303 I've added a note about this into the array on the latest version, but would appreciate you taking a look as your comment is on an outdated version.

---
date: 2022-10-31
title: "Array 1.41.0: SUBHEADLINE GOES HERE"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
categories: ['Release notes', 'Product updates']
featuredImage: ../images/blog/posthog-array-blog.png
featuredImageType: full
---

Want to know more about what we're up to? [Subscribe to HogMail, our newsletter](/newsletter), which we send every two weeks!

> Running a self-hosted instance? Check out our [guide to upgrading PostHog](/docs/runbook/upgrading-posthog).

## PostHog 1.41.0 release notes

**Release highlights:**

- New: Count of events per user
- Improved: Math selector
- New: Text cards on dashboards
- Improved: Recordings interface

> - New: Plugin metrics

Reviewer 2: ```suggestion
- New: App metrics
```

- New: Persons on events on by default
- New: Alias changes
- New: Ingestion warnings
- New: Groups caching changes
- Finally: Change your email yourself
- New: Hedgehog mode
- New: github subdirectories in apps
- One more thing: Site apps 

### New: Count of events per user

PICTURE GOES HERE

*Average number of purchases per user? Maximum number of forms submitted per user?* 
Questions like these used to be hard to answer, but no more! Introducing "count of events per user".

### Improved: Math selector

PICTURE GOES HERE

Work on the previous task had us redo the math selector.

### New: Text cards on dashboards

PICTURE GOES HERE

Would you like to add some metadata or clarifications on your dashboards? Now you can! Hint: gifs are metadata too! 

### Improved: Recordings interface

PICTURE GOES HERE

We've been listening and implementing the most requested features, with a special focus on sparking joy. 
It's like a completely new app now, so check it out! 

> ### New: Plugin metrics

Reviewer 2: ```suggestion
### New: App metrics
```

PICTURE GOES HERE

> Curious how well your plugins are doing? You no longer have to read smoke signals in AWS logs - just use the new Plugin Metrics page. 

Reviewer 2: TODO @macobo .

Notes on what kafka topics to add as well

### New: Persons on events on by default

PICTURE GOES HERE

> We used to store events in one table and persons in another. Any query that touched person properties would require 

Reviewer 2: This is all inaccurate. @yakkomajuri you have the latest context on async migration + it being turned on automatically, will add notes on queries after that.

database level JOINs and time out, once you reached the billion event scale. No longer! By moving person data
onto the events themselves, we eliminated a whole class of performance issues 🚀.

### New: Alias changes

PICTURE GOES HERE

The above change required us to be strict about who's a identified user and who's not. Because of that, you might
benefit from double checking your `identify` and `alias` calls. The new rule is this: only anonymous users can be
aliased into other users. Once you call `identify` on a user, you can no longer alias this user into another identified
user. 

### New: Ingestion warnings

PICTURE GOES HERE

> In case you still try to alias identified events, there's a new Ingestion Warnings page to remind you that it's a bad idea. 

Reviewer 2: TODO @tiina303 .

Notes on what kafka topics to add as well

Reviewer: Was this something that needed to happen manually, I can't recall 🤷‍♀️  Maybe you can provide the guidelines about that?

> Curious how well your apps are doing? Previously, you may have had to pour over the AWS logs, but now you can head to the new app metrics page to find out how many events an app has processed, how many retries were attempted and what errors may have occured. Very handy. 

Reviewer 2: ```suggestion
Curious how well your apps are doing? Previously, you may have had to pour over the AWS logs, but now you can head to the new app metrics page to find out how many events an app has processed and what errors may have occurred. Very handy. 
```

Reviewer 2: cc @joethreepwood does the suggestion above make sense wording-wise? This feature is _not_ generally available.

Author: ```suggestion
Curious how well your apps are doing? Previously, you may have had to pour over the AWS logs, but now you can head to the new app metrics page to find out how many events an app has processed, how many retries were attempted and what errors may have occured. Very handy. 

App metrics are only available for users on Scale or Enterprise plans. 
```

Questions like these used to be hard to answer with PostHog, but no more! We've added a new series option for insights, enabling you to see the count of events per user - including sub-options for seeing the averages, minimum and maximum count of events per user. 

**Bonus update:** While working on this feature, we also took the opportunity to improve our math selector UI.

### New: Text cards on dashboards

![text cards on dashboards](../images/blog/array/1-41-0-text-card-on-dashboard.png)

Previously there was no easy way to add context or links to a dashboard, meaning you may have to send long explanations when sharing a dashboard with teammates. That's why we've added the option for users on paid plans to add text cards where they can add any information they want, including metadata, images or gifs!

### New: Ingestion warnings

![ingestion warnings](../images/blog/array/1-41-0-ingestion.png)

We've added a new page to the Data Management section which lists warnings related to data ingestion from the past 30 days. If you still try to merge identified users into others, the Ingestion Warning page is where we'll remind you that the merge got blocked.

> > **Note:** Self-hosted users managing kafka separately should create a new topic `clickhouse_ingestion_warnings` manually.

Reviewer 2: @joethreepwood A note on what some self-hosted users may need to do for this to work. Feel free to update this as you'd like.

### New: App metrics

![app metrics](../images/blog/array/1-41-0-app-metrics-insight.png)

Curious how well your apps are doing? Previously, you may have had to pour over the AWS logs, but now you can head to the new app metrics page to find out how many events an app has processed, how many retries were attempted and what errors may have occured. Very handy. 

Want to take a look? Head to the apps page in your instance and click the chart symbol for any installed app.  

> > **Note:** Self-hosted users managing kafka separately should create a new topic `clickhouse_app_metrics` manually.

Reviewer 2: @joethreepwood A note on what some self-hosted users may need to do for this to work. Feel free to update this as you'd like.

### Other improvements & fixes

You think that's it? Not by a long shot! Version 1.41 also adds hundreds of other improvements and fixes, including...

- **Improvement:**  You can now [send analytics events from GitHub actions, to PostHog](https://github.com/PostHog/posthog-github-action)
- **Improvement:**  We have revamped our timezone system! We've squashed various bugs and improved the interval grouping to be more in line with expectations when filtering on dates.
- **Fix:** WAU/MAU aggregation in Trends was quietly always grouped by day, even if a different interval (hour/week/month) was selected. Additionally, those modes showed zero users for periods with no relevant events, even if the real count should have been non-zero due to WAU/MAU being a trailing count. Both issues are now fixed, and we've expanded our test coverage of those aggregation modes to ensure their results are accurate going forward.
- **Fix**: Experiment results will appear immediately after the first exposure to a user

> - **Fix**: The experiments table is now sortable

Author: ```suggestion
- **Improvement**: The experiments table is now sortable
```

You won’t see any UI changes as a result of this change — persons will still have their own Persons & Groups section on the sidebar, for example — but you will notice results are a lot (up to 400%!) faster for any queries involving persons, groups or events. This is a massive change, so be sure to [read the full announcement for more info](/blog/persons-on-events).

> **Note:** As a result of putting persons on events we now have to be strict about identified and unidentified users. Now, only anonymous users can be aliased into other users - meaning that once you call `identify` on a user, you can no longer alias the user into another identified user. Because of that, we recommend double checking your `identify` and `alias` calls, and following advice above [how to upgrade a self-hosted instance to 1.41.0](/blog/the-posthog-array-1-41-0). 

### New: Count of events per user

![count events per user](../images/blog/array/1-41-0-count-events-per-user.png)

Have you ever asked yourself '*What's the average average number of purchases per user?*' or '*What's the maximum number of forms submitted per user?'* 

> Questions like these used to be hard to answer with PostHog, but no more! We've added a new series option for insights, enabling you to see the count of events per user - including sub-options for seeing the averages, minimum and maximum count of events per user. 

Reviewer 3: I think the most important goal that can now be achieved is _measuring intensity_. Also, if we're listing functions, we can just mention all of them since that's not an unwieldy long list:
```suggestion
Questions like these used to be hard to answer with PostHog, but no more! Use the new "Count per user" aggregation mode available in Trends to analyze _how intensely_ your users use the product and its features. "Count per user" supports common statistical functions for crunching the per-user numbers: average, median, minimum, maximum, 90th/95th/99th percentile.
```

> **Bonus update:** While working on this feature, we also took the opportunity to improve our math selector UI.

Reviewer 3: ```suggestion
**Bonus:** While working on this feature, we also took the opportunity to improve our aggregation selector UI: aggregation by property value is now presented with more clarity.
```

### New: Text cards on dashboards

![text cards on dashboards](../images/blog/array/1-41-0-text-card-on-dashboard.png)

Previously there was no easy way to add context or links to a dashboard, meaning you may have to send long explanations when sharing a dashboard with teammates. That's why we've added the option for users on paid plans to add text cards where they can add any information they want, including metadata, images or gifs!

### New: Ingestion warnings

![ingestion warnings](../images/blog/array/1-41-0-ingestion.png)

We've added a new page to the Data Management section which lists warnings related to data ingestion from the past 30 days. If you still try to merge identified users into others, the Ingestion Warning page is where we'll remind you that the merge got blocked.

> **Note:** Self-hosted users managing kafka separately should create a new topic `clickhouse_ingestion_warnings` manually.

> 

Author: ```suggestion
```

### One more thing: Site apps 

![site-apps](../images/blog/array/1-41-0-site-apps.gif)

We're testing a new big (beta) thing: site apps. You need to **manually opt in** to enable this feature by configuring your `posthog-js` initialization to include `opt_in_site_apps: true`. Once you do, PostHog will be able to inject code onto your website through `posthog-js`. We've put together [a tutorial that explains how to make a site app](/tutorials/build-site-app) if you're interested. 

Site apps can be useful for a number of potential tasks, such as displaying feedback forms, posting service update banners, or making it rain pineapples. No, we're not really sure why we made the last one either. 

### Other improvements & fixes

You think that's it? Not by a long shot! Version 1.41 also adds hundreds of other improvements and fixes, including...

- **Improvement:**  You can now [send analytics events from GitHub actions, to PostHog](https://github.com/PostHog/posthog-github-action)
- **Improvement:**  We have revamped our timezone system! We've squashed various bugs and improved the interval grouping to be more in line with expectations when filtering on dates.

> - **Improvement:** 

Reviewer 3: Not sure if something else should have been here, but I think we should mention https://github.com/PostHog/posthog/pull/12302:
```suggestion
- **Fix:** WAU/MAU aggregation in Trends was quietly always grouped by day, even if a different interval (hour/week/month) was selected. Additionally, those modes showed zero users for periods with no relevant events, even if the real count should have been non-zero due to WAU/MAU being a trailing count. Both issues are now fixed, and we've expanded our test coverage of those aggregation modes to ensure their results are accurate going forward.

- **Fix**: Experiment results will appear immediately after the first exposure to a user

> - **Fix**: The experiments table is now sortable

Reviewer 3: This sounds more like an improvement, rather than a fix? Unless the table wrongly indicated it was sortable, but this could be clearer

---
date: 2022-11-02
title: "Array 1.41.0: Improving performance by up to 400%"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
categories: ['Release notes', 'Product updates']
featuredImage: ../images/blog/posthog-array-blog.png
featuredImageType: full
---

Want to know more about what we're up to? [Subscribe to HogMail, our newsletter](/newsletter), which we send every two weeks!

## 1.41.0 update guide for self-hosted users

> Due to changes in this update, it is important check `alias` usage before [upgrading PostHog](/docs/runbook/upgrading-posthog) to 1.41.0 on a self-hosted instance. For example, assuming `email` is used as the identified user id, then: 

Reviewer: might be nice to link to https://posthog.com/docs/integrate/identifying-users#considerations e.g as "read more"

Author: ```suggestion
Due to changes in this update, it is important check `alias` usage before [upgrading PostHog](/docs/runbook/upgrading-posthog) to 1.41.0 on a self-hosted instance. Further [information is available in the docs](/docs/integrate/identifying-users#considerations), but for example, assuming `email` is used as the identified user id, then: 
```

```

> identify(email)                                   # in the frontend
> alias(backend_unique_id, email)      # in the backend - THIS WILL NOT WORK
> alias(email, backend_unique_id2)    # in the backend - the right order

Reviewer 4: ```suggestion
identify(email)                     # in the frontend
alias(email, backend_unique_id)     # in the backend - this works
alias(backend_unique_id, email)     # in the backend - THIS WILL NOT WORK
```

```

> > **If you haven't run async migrations 0005, 0006 and/or 0007 before:** Before updating to 1.41.0, check [ingestion warnings](#new-ingestion-warnings) and solve any outstanding issues. Run async migrations 0005, 0006 and/or 0007 at `<your-posthog-site>/instance/async_migrations`. 

Reviewer: This is important, they shouldn't run the async migrations on the old version & we only care about 0007 being completed of not, but they might have 0005 and/or 0006 in either state.
```suggestion
> **If you haven't run async migration 0007 before:** Before updating to 1.41.0, check [ingestion warnings](#new-ingestion-warnings) and solve any outstanding issues. Then upgrade to 1.41.0 and then run async migration 0007 at `<your-posthog-site>/instance/async_migrations` (if you haven't ran 0005 yet start with that, then 0006 and finally 0007). 
```

> > **If you have completed async migration 0007 before:** If you have changed anything regarding `alias` or based on ingestion warnings, you must re-run async migration 0007 by [connecting to Postgres](/docs/self-host/deploy/troubleshooting#how-do-i-connect-to-postgres), running `UPDATE posthog_asyncmigration SET status = 0 WHERE name = '0007_persons_and_groups_on_events_backfill' AND status = 2;` and re-running 0007 at `<your-posthog-site>/instance/async_migrations`.

Reviewer: these are both crucial details:
```suggestion
> **If you have completed async migration 0007 before:** If you have changed anything regarding `alias` or based on ingestion warnings or if you use groups, you must re-run async migration 0007 on top of 1.41.0 by [connecting to Postgres](/docs/self-host/deploy/troubleshooting#how-do-i-connect-to-postgres), running `UPDATE posthog_asyncmigration SET status = 0 WHERE name = '0007_persons_and_groups_on_events_backfill' AND status = 2;` and re-running 0007 at `<your-posthog-site>/instance/async_migrations`.
```

### New: Ingestion warnings

![ingestion warnings](../images/blog/array/1-41-0-ingestion.png)

We've added a new page to the Data Management section which lists warnings related to data ingestion from the past 30 days. If you still try to merge identified users into others, the Ingestion Warning page is where we'll remind you that the merge got blocked.

> **Note:** Self-hosted users managing kafka separately should create a new topic `clickhouse_ingestion_warnings` manually.

### New: App metrics

![app metrics](../images/blog/array/1-41-0-app-metrics-insight.png)

Curious how well your apps are doing? Previously, you may have had to pour over the AWS logs, but now you can head to the new app metrics page to find out how many events an app has processed, how many retries were attempted and what errors may have occured. Very handy. Want to take a look? Head to the apps page in your instance and click the chart symbol for any installed app.  

App metrics are only available for users on Scale or Enterprise plans. 

> **Note:** Self-hosted users managing kafka separately should create a new topic `clickhouse_app_metrics` manually.

> ### Improved: Recordings interface

Reviewer 5: This is a huge deal, and we should probably have this above most of the other updates (certainly above the text in dashboards)

Reviewer 4: lol, I already [moved it up](https://github.com/PostHog/posthog.com/pull/4512/commits/275b292ee11b335021ad20136bd20ee77d935692) by 4 places. Could be even more.
